// gitHubIssue/lib/resources.js
"use strict";
module.exports = {
	headers: {
		Accept: "application/vnd.github.v3+json",
		"User-Agent": "GitHubIssues/0.1 VisualStudioCode/0.10",
		"Connection": "keep-alive"
	},
	hostname: "api.github.com",
	strings: {
		"400": "Invalid request submitted",
		"401": "Credentials invalid",
		"403": "Authentication required",
		"404": "Possible authentication required",
		"422": "Invalid fields submitted",
		redirectError: "Too many redirects, bailed",
		unknown: "An unknown error occured"
	},
	icons: ["alert", "arrow-down", "arrow-left", "arrow-right", "arrow-small-down", "arrow-small-left",
		"arrow-small-right", "arrow-small-up", "arrow-up", "beaker", "bell", "bold", "book", "bookmark", "briefcase",
		"broadcast", "browser", "bug", "calendar", "check", "checklist", "chevron-down", "chevron-left", "chevron-right",
		"chevron-up", "circle-slash", "circuit-board", "clippy", "clock", "cloud-download", "cloud-upload", "code",
		"color-mode", "comment-discussion", "comment", "credit-card", "dash", "dashboard", "database", "desktop-download",
		"device-camera-video", "device-camera", "device-desktop", "device-mobile", "diff-added", "diff-ignored",
		"diff-modified", "diff-removed", "diff-renamed", "diff", "ellipsis", "eye", "file-binary", "file-code",
		"file-directory", "file-media", "file-pdf", "file-submodule", "file-symlink-directory", "file-symlink-file",
		"file-text", "file-zip", "flame", "fold", "gear", "gift", "gist-secret", "gist", "git-branch", "git-commit",
		"git-compare", "git-merge", "git-pull-request", "globe", "graph", "heart", "history", "home", "horizontal-rule",
		"hubot", "inbox", "info", "issue-closed", "issue-opened", "issue-reopened", "italic", "jersey", "key", "keyboard",
		"law", "light-bulb", "link-external", "link", "list-ordered", "list-unordered", "location", "lock", "logo-gist",
		"logo-github", "mail-read", "mail-reply", "mail", "mark-github", "markdown", "megaphone", "mention", "milestone",
		"mirror", "mortar-board", "mute", "no-newline", "octoface", "organization", "package", "paintcan", "pencil",
		"person", "pin", "plug", "plus", "primitive-dot", "primitive-square", "pulse", "question", "quote", "radio-tower",
		"repo-clone", "repo-force-push", "repo-forked", "repo-pull", "repo-push", "repo", "rocket", "rss", "ruby", "search",
		"server", "settings", "shield", "sign-in", "sign-out", "squirrel", "star", "stop", "sync", "tag", "tasklist",
		"telescope", "terminal", "text-size", "three-bars", "thumbsdown", "thumbsup", "tools", "trashcan", "triangle-down",
		"triangle-left", "triangle-right", "triangle-up", "unfold", "unmute", "versions", "watch", "x", "zap"],
	style: ".label{display:inline-block;padding:5px 10px;border-radius:5px;color:white}"
};
